src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs

```diff
@@ -35,54 +38,90 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
+
+                using var _ = ArrayBuilder<VSCodeAction>.GetInstance(out var results);
+                foreach (var codeAction in codeActions)
+                {
+                    // Temporarily filter out suppress and configure code actions, as we'll later combine them under a top-level
+                    // code action.
+                    if (codeAction.Key is AbstractConfigurationActionWithNestedActions)
                     {
-                        CommandIdentifier = RunCodeActionCommandName,
-                        Title = codeAction.Title,
-                        Arguments = new object[]
-                        {
-                                new RunCodeActionParams
-                                {
-                                    CodeActionParams = request,
-                                    Title = codeAction.Title
-                                }
-                        }
-                    });
+                        continue;
+                    }
+
+                    results.Add(GenerateVSCodeAction(request, codeAction.Key, codeAction.Value));
+                }
+
+                // Special case (also dealt with specially in local Roslyn): 
+                // If we have configure/suppress code actions, combine them under one top-level code action.
+                var configureSuppressActions = codeActions.Where(a => a.Key is AbstractConfigurationActionWithNestedActions);
+                if (configureSuppressActions.Any())
+                {
+                    results.Add(GenerateVSCodeAction(request, new CodeActionWithNestedActions(
+                        CodeFixesResources.Suppress_or_Configure_issues,
+                        configureSuppressActions.Select(a => a.Key).ToImmutableArray(), true), CodeActionKind.QuickFix));
+                }
+
+                return results.ToArray();
             }
 
-            return result.ToArrayAndFree();
+            static VSCodeAction GenerateVSCodeAction(
+                CodeActionParams request,
+                CodeAction codeAction,
+                CodeActionKind codeActionKind,
+                string parentTitle = "")
+            {
+                using var _ = ArrayBuilder<VSCodeAction>.GetInstance(out var nestedActions);
+                foreach (var action in codeAction.NestedCodeActions)
+                {
+                    nestedActions.Add(GenerateVSCodeAction(request, action, codeActionKind, codeAction.Title));
+                }
+
+                return new VSCodeAction
+                {
+                    Title = codeAction.Title,
+                    Kind = codeActionKind,
+                    Diagnostics = request.Context.Diagnostics,
+                    Children = nestedActions.ToArray(),
+                    Data = new CodeActionResolveData { CodeActionParams = request, DistinctTitle = parentTitle + codeAction.Title }
```

Comment:
are there cases where the nested code action titles conflict between two different nested groups?  If so, I wonder if we could use the codeactionprovider type name + title instead of having to specify the entire tree (unless the title can be the same within a nested group)

we should also doc on the CodeActionResolveData what goes inside the DistinctTitle

Reply:
Hmm, I think this would fail with 'Suppress or Configure issues' if we had two diagnostics that we could suppress/configure, e.g. we would have "ConfigureSeverityCodeFixProvider" + "None" twice in this case.